### DISCUSSION TITLE
Composite actions unable to access vars context

### DISCUSSION THREAD

**Discussion by ilyamt-tandemg** [community/community#43878] [Actions]
### Select Topic Area

Bug

### Body

The `vars` context was used in a composite action executed on a self-hosted runner. The bug for confirmed on both Linux and Windows.

> Error: /home/***/action-runners/worker/_work/***/./.github/actions/***/action.yml (Line: 39, Col: 12): Unrecognized named-value: 'vars'. Located at position 1 within expression: vars.MY_VAR

The self-hosted runner's version is the currently latest: 
> Current runner version: '2.299.1'

**Comment by Simran-B** [comment]
Please share a code snippet that shows how and where you use `vars.MY_VAR`

**Comment by ilyamt-tandemg** [comment]
This is the workflow being called first:
```
name: vars context test

on:
  push:
    branches:
      - tmp_githubactions

jobs:
  job1:
    runs-on: self-hosted
    steps:
      - uses: actions/checkout@v3
      - run: |
          echo ${{ vars.MY_VAR }}
          echo ${{ vars.ARTIFACTS_S3_BUCKET_NAME }}
      - uses: ./.github/actions/embedded-cmake-configure
  job2:
    uses: ./.github/workflows/hello.yml
```

The workflow itself is able to access `vars`, and also the reusable workflow `hello.yml`. The problem is only with the composite action `embedded-cmake-configure`, which looks exactly as the workflows:
```
name: Test action
description: .

runs:
  using: "composite"
  steps:
    - shell: ${{ runner.os == 'Windows' && 'cmd' || 'bash' }}
      run: |
        echo ${{ vars.MY_VAR }}
        echo ${{ vars.ARTIFACTS_S3_BUCKET_NAME }}
```

Hence, I conclude that the problem is limited to accessing `vars` from composite actions. It looks as if the template for the composite actions was not updated properly:

```Error: c:\***\./.github/actions/embedded-cmake-configure\action.yml (Line: 8, Col: 12):
Error: c:\***\./.github/actions/embedded-cmake-configure\action.yml (Line: 8, Col: 12): Unrecognized named-value: 'vars'. Located at position 1 within expression: vars.MY_VAR
Error: GitHub.DistributedTask.ObjectTemplating.TemplateValidationException: The template is not valid. c:\***\./.github/actions/embedded-cmake-configure\action.yml (Line: 8, Col: 12): Unrecognized named-value: 'vars'. Located at position 1 within expression: vars.MY_VAR
   at GitHub.DistributedTask.ObjectTemplating.TemplateValidationErrors.Check()
   at GitHub.Runner.Worker.ActionManifestManager.ConvertRuns(IExecutionContext executionContext, TemplateContext templateContext, TemplateToken inputsToken, String fileRelativePath, MappingToken outputs)
   at GitHub.Runner.Worker.ActionManifestManager.Load(IExecutionContext executionContext, String manifestFile)
Error: Fail to load c:\***\./.github/actions/embedded-cmake-configure\action.yml
```

**Comment by ChristopherHX** [comment]
A cite from an actions/runner developer

> +1 we dont need in actions yaml file, since we support vars only in workflow file.

source: https://github.com/actions/runner/pull/2096#discussion_r976597738

The initial PR would have allowed your usecase, but it has been reverted during review.

**Reply by ChristopherHX** [reply]
You could add all `vars` as `inputs` of your composite action
```yaml
- uses: ./
  with: ${{ vars }}
- uses: ./
  with:
    myinput: xf
    ${{ insert }}: ${{ vars }}
```

Then just use `${{ inputs.varname }}` in your composite action

**Reply by ilyamt-tandemg** [reply]
> You could add all vars as inputs of your composite action
> The initial PR would have allowed your usecase, but it has been reverted during review.

- https://docs.github.com/en/actions/learn-github-actions/contexts#vars-context
This is not mentioned in the docs.
- https://docs.github.com/en/actions/learn-github-actions/contexts#secrets-context
This does not follow the rule of least surprise as the idea for excluding `secrets` in composite actions is security; `vars` are by definition security-cleared-secrets, so the argument does not apply. 
- The trivial use for `vars` is configuring something directly -- without making it the responsibility of the caller. Any boilerplate solution like the one proposed above will increase boilerplate and force the caller to know the implementation of the callee.
- Similar limitations and their workarounds are sort of "arcane knowledge" that prevents me from asking a random developer to take responsibility for any part of the CI without being his or her nanny.

Thanks for the proposed workaround. We already use something similar based on `env` to be explicit about the params that we want to pass and the params we are forced to pass. But given the reasons above we are not sure that this should be classified as a solution.
